src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AssigningSymbolAndItsMemberInSameStatement.Fixer.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = AssigningSymbolAndItsMemberInSameStatement.RuleId), Shared]
+    public class AssigningSymbolAndItsMemberInSameStatementFixer : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(AssigningSymbolAndItsMemberInSameStatement.RuleId);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            SyntaxNode root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            SyntaxNode node = root.FindNode(context.Span);
+
+            if (node == null)
+            {
+                return;
+            }
+
+            string title = MicrosoftCodeQualityAnalyzersResources.AssigningSymbolAndItsMemberInSameStatementTitle;
+            context.RegisterCodeFix(new MyCodeAction(title,
+                 async ct => await SplitAssignment(context.Document, node, ct).ConfigureAwait(false),
+                 equivalenceKey: title),
+            context.Diagnostics);
+        }
+
+        private static async Task<Document> SplitAssignment(Document document, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            DocumentEditor editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+            var parentAssignment = (Microsoft.CodeAnalysis.CSharp.Syntax.AssignmentExpressionSyntax)node.Parent;
+
+            var rightAssignment = (Microsoft.CodeAnalysis.CSharp.Syntax.AssignmentExpressionSyntax)parentAssignment.Right.WithTriviaFrom(parentAssignment);
+
+            var newSecondAssignment = parentAssignment.WithRight(rightAssignment.Right);
+
+            editor.ReplaceNode(parentAssignment, rightAssignment);
+            editor.InsertAfter(rightAssignment, newSecondAssignment);
```

Comment:
I'm experimenting to implement the codefix, but it's not working. I believe the problem is making multiple modifications to the DocumentEditor, but I don't know how to get it working.

Note: I know that my current implementation here need to be moved to a `CSharpAssigningSymbolAndItsMemberInSameStatement.Fixer`, and make a parallel one for VB, and also mark the parent as abstract. I'll do that later when I get a first working code here.

@mavasani Can you help here? Is there something I should write in between the `ReplaceNode` and `InsertAfter` node? or there is another approach to do what I want?

Reply:
I am not an expert (yet) with code-fixes but there is something I don't get with the logic you are putting in place here. In your example `a.Field = a = b` we have:
- `parentAssignment` = `a.Field = a = b`
- `rightAssignment` = `a = b`
- `newSecondAssignment` = `a.Field = b`

Now you replace `a.Field = a = b` with `a = b` and then you add right after `a.Field = b` but in the expected code-fix you have written `a.Field = b = a`.

IMO (note that I haven't tested) we should generate a new node of type equality (using SyntaxGenerator) with left `b` and right `a` taken from `rightAssignment` and then replace the code of `newSecondAssignment` to be `var newSecondAssignment = parentAssignment.WithRight(newEqualityNode)`. After that all you need to do is the replace.

Reply:
I couldn't understand the reasoning for my code not to work.

But let me try the path you referred to.

The method I find that best suite the situation in the SyntaxFactory class is `SyntaxFactory.List(IEnumerable<TNode>)`. (Did you mean that method?)

So, I give it the `rightAssignment` and the `newSecondAssignment`, it will return `SyntaxList<TNode>`. Now, what should I do with that returned value?, how to replace the parentAssignment with the returned `SyntaxList`?

Reply:
> I am not an expert (yet) with code-fixes

You're really much better than me :)
I'm just a beginner, trying to gain some experience with open-source projects.

Reply:
Sorry I was too lightweight on my explanation. The `SyntaxFactory` allows you to create any new kind of syntax you could find in C#. In your example, you want to transform `a.Field = a = b` into `a.Field = b = a`, to do so you actually want to create a new assignment expression (`b = a`) out of the old assignment expression `a = b`). You should be able to do something like `SyntaxFactory.AssignmentExpression(SyntaxKind.SimpleAssignmentExpression, rightAssignment.Right, rightAssignment.Left)` which will return the new assignment. Then if you do the `parentAssignment.WithRight(reversedAssignment)` you will replace the right part of `a.Field = a = b` (i.e. `a = b`) with your new assignment `b = a` and so end-up with `a.Field = b = a`. 

Note that you will probably have to add `WithoutTrivia` on the new right part to avoid the space before the semicolon `b = a ;` and you might also have to add some trailing space on the new left part to avoid `b= a ;`.

I would also suggest to add one more test with `a.Field = a = b = c;` to ensure the code is working properly for a more complex scenario.

And also you might want to use safe-cast rather than cast to avoid having your codefix blowing in case of invalid code. It's possible to have invalid (not compiling) code.

Reply:
Ops, Sorry, My fault. The test I wrote isn't correct.
I want to transform `a.Field = a = b;` into **two statements**:

```csharp
a = b;
a.Field = b;
```

Reply:
Although the explanation was for the unintended behavior stated in my wrong test, but the explanation was useful anyway :)

Reply:
Oh! I see! Hum I will need to pull your PR and make some tests because I don't know enough codefixes to see what could be wrong.

Reply:
Looking at the test result, it seems that the new node is not part of a syntax list as requested by the [InsertAfter method](https://docs.microsoft.com/dotnet/api/microsoft.codeanalysis.editing.syntaxeditor.insertafter?view=roslyn-dotnet#Microsoft_CodeAnalysis_Editing_SyntaxEditor_InsertAfter_Microsoft_CodeAnalysis_SyntaxNode_Microsoft_CodeAnalysis_SyntaxNode_). I am not sure what shall be called to add a new instruction to the blocks so I will do a bit of debugging.

Reply:
I have found something that seems to be working but it feels quite wrong so I suspect there is a better way. My solution:

1. Create 2 new `ExpressionStatementSyntax` using the `SyntaxFactory` to represent the 2 statements you want to add.
2. Find ancestor of type `BlockSyntax`
3. Take all statements of this block and replace the old statement with the 2 new
4. Call `WithStatements`
5. Call `ReplaceNode` to replace the block

@sharwell Any pointer on a better solution?

Reply:
@Evangelink Thanks for making some trials and getting to that solution.
I'll wait and see if @sharwell sees a better solution.

Reply:
ping @sharwell

Reply:
@mavasani If you have the chance to ping @sharwell in a meeting, that would help @Youssef1313 moving forward on this PR.